src/mysql_conn.erl

```diff
@@ -91,54 +92,109 @@ init(Opts) ->
         N when N > 0 -> N
     end,
 
+    State0 = #state{
+        tcp_opts = TcpOpts,
+        ssl_opts = SSLOpts,
+        host = Host, port = Port,
+        user = User, password = Password,
+        database = Database,
+        queries = Queries, prepares = Prepares,
+        log_warnings = LogWarn,
+        connect_timeout = ConnectTimeout,
+        ping_timeout = PingTimeout,
+        query_timeout = QueryTimeout,
+        query_cache_time = QueryCacheTime,
+        cap_found_rows = (SetFoundRows =:= true)
+    },
+
+    case proplists:get_value(connect_mode, Opts, synchronous) of
+        synchronous ->
+            case do_connect(State0) of
+                {ok, State1} ->
+                    {ok, State1};
+                {error, Reason} ->
+                    {stop, Reason}
+            end;
+        asynchronous ->
+            gen_server:cast(self(), connect),
+            {ok, State0};
+        lazy ->
+            {ok, State0}
+    end.
+
+do_connect(#state{connect_timeout = ConnectTimeout} = State) ->
+    Self = self(),
+    Pid = spawn_link(
+        fun () ->
+            case do_connect1(State) of
+                {ok, #state{sockmod = SockMod, socket = Socket} = State1} ->
+                    SockMod:controlling_process(Socket, Self),
+                    Self ! {self(), {ok, State1}};
```

Comment:
The Self variable could be called MainPid or something. Self is not self everywhere, so it's not a very good name.

Reply:
Another bad habit of mine, will change :)

Reply:
Changed as suggested in latest commit.